### PR TITLE
docs: clarify http header api doc

### PR DESCRIPTION
**Pull request by jbogarthyde**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
HTTP Header API doc does not make it obvious that changes are made to a clone of the original instance.

Issue Number: https://github.com/angular/angular/issues/33033

## What is the new behavior?
Clarify API docs 
https://angular.io/api/common/http/HttpHeaders

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/http/src/headers.ts [hunk 1]
```diff
@@ -85,11 +85,11 @@ export class HttpHeaders {
   }
 
   /**
-   * Checks for existence of a header by a given name.
+   * Checks for existence of a given key in the header.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: I think the previous terminology is more correct. The new descriptions makes it feel like the full HTTP block of headers is just referred to as "header", whereas a "header" typically only refers to a single "header" line:

For instance, the HTTP snippet:

```http
POST /api/v1/users
Host: https://example.com
Content-Type: application/json
Accept: application/json
```

has three headers, named `Host`, `Content-Type` and `Accept`. I think the term "name" is more appropriate to use compared to "key", as a key implies that its occurrence is unique. This is not a necessity with HTTP, where the following is allowed:

```http
POST /api/v1/users
Host: https://example.com
Content-Type: application/json
Accept: application/json
Accept: text/plain
```

This HTTP request will be interpreted to have its `Accept` header set to `application/json, text/plain`.

File: packages/common/http/src/headers.ts [hunk 2]
```diff
@@ -148,24 +149,26 @@ export class HttpHeaders {
...
+   * @param value The value or values to set or overide for the given key.
    *
    * @returns A clone of the HTTP header object with the newly set header value.
    */
   set(name: string, value: string|string[]): HttpHeaders {
     return this.clone({name, value, op: 's'});
   }
   /**
-   * Deletes all header values for a given name.
+   * Deletes header values for a given key name.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: Could we make it explicit that the original object is not modified and that the change can only be observed in the returned object? Having it only mentioned in the `@returns` is not so visible, especially on AIO where the main description is at the top of the method's card, and the return type description is all the way at the bottom:

image

When reading just the method description (which is typically the case!), it is very natural to assume that the original object is updated.

This holds true for all methods that return a cloned object.

File: packages/common/http/src/headers.ts [hunk 3]
```diff
@@ -14,8 +14,8 @@ interface Update {
 
 /**
  * Represents the header configuration options for an HTTP request.
- *
- * Instances should be assumed immutable with lazy parsing.
+ * Instances are immutable with lazy parsing.
+ * Modifying methods return a cloned instance with the change.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: Could we add somehow that the original object is not changed?

```suggestion
 * Modifying methods return a cloned instance with the change, the original object is never changed.
```

File: packages/common/http/src/headers.ts [hunk 4]
```diff
@@ -14,8 +14,8 @@ interface Update {
 
 /**
  * Represents the header configuration options for an HTTP request.
- *
- * Instances should be assumed immutable with lazy parsing.
+ * Instances are immutable with lazy parsing.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: To me this (as well as the original) sounds very abstract. To me, the concept of "parsing" is strange in the context of keeping track of a collection of headers. I would prefer if the mentioning of the object being immutable is fused with the subsequent sentence a bit more; i.e. what does it mean for this object to be immutable.

**Review by JoostK — CHANGES_REQUESTED** [review]
Thanks for taking this on! I left some comments inline.

Also, could the descriptions be updated to make use of the 100 chars in width? It feels strange that they are kept so short in the code.

File: packages/common/http/src/headers.ts (hunk 4 above)
Comment (jbogarthyde) [reply packages/common/http/src/headers.ts]: Done. (I didn't understand that sentence at all, so I left it unchanged. Can suggest a clearer way to say it, or can I just take it out?)

File: packages/common/http/src/headers.ts (hunk 1 above)
Comment (jbogarthyde) [reply packages/common/http/src/headers.ts]: Thanks for the clarification - -removing the "key" concept.

File: packages/common/http/src/headers.ts (hunk 4 above)
Comment (JoostK) [reply packages/common/http/src/headers.ts]: I would be fine with taking the statement about "lazy parsing" out :+1: It's an implementation detail that does not affect the observable behavior, so my vote would be to remove it.

File: packages/common/http/src/headers.ts [hunk 5]
```diff
@@ -136,36 +136,39 @@ export class HttpHeaders {
   }
 
   /**
-   * Appends a new header value to the existing set of
-   * header values.
+   * Appends a new value to the existing set of
+   * values for a header and returns a clone of the original headers instance.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: This sentence allows for the interpretation of the appending to also occur in the original instance (before the cloning). I prefer the way the `set` method reads "in a clone of the original instance".

Also, could the first line here span the available width and any overflow follow on the second line? The current spacing is quite awkward to read.

File: packages/common/http/src/headers.ts [hunk 6]
```diff
@@ -136,36 +136,39 @@ export class HttpHeaders {
...
   append(name: string, value: string|string[]): HttpHeaders {
     return this.clone({name, value, op: 'a'});
   }
   /**
-   * Sets a header value for a given name. If the header name already exists,
-   * its value is replaced with the given value.
+   * Sets or modifies a key value for a given header
+   * in a clone of the original instance.
+   * If the header already exists,
+   * its value is replaced with the given value in the returned object.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: nit: could the spacing here also be cleaned up a bit?

File: packages/common/http/src/headers.ts [hunk 7]
```diff
@@ -123,9 +123,9 @@ export class HttpHeaders {
   }
 
   /**
-   * Retrieves a list of header values for a given header name.
+   * Retrieves a list of header values for a given name.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: Could this be made consistent with `has` and `get`, which both use "for a given header". I think I like that better 😄

File: packages/common/http/src/headers.ts [hunk 8]
```diff
@@ -98,11 +98,11 @@ export class HttpHeaders {
   }
 
   /**
-   * Retrieves the first header value that matches a given name.
+   * Retrieves the first value of a given header.
    *
-   * @param name The header name to retrieve.
+   * @param name The header name to query.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: I prefer the original 😇

File: packages/common/http/src/headers.ts [hunk 9]
```diff
@@ -85,11 +85,11 @@ export class HttpHeaders {
   }
 
   /**
-   * Checks for existence of a header by a given name.
+   * Checks for existence of a given header.
    *
-   * @param name The header name to check for existence.
+   * @param name The name to check for existence.
```
Comment (JoostK) [inline packages/common/http/src/headers.ts]: I prefer the original :)

**Review by JoostK — APPROVED** [review]

**Comment by jbogarthyde** [comment]
Caretaker: Docs only

**Review by alxhub — APPROVED** [review]

**Pull request closed without merging**